packages/docker-engine/0001-read-default-oci-spec-from-file.patch

```diff
@@ -0,0 +1,170 @@
++++ b/daemon/config/config.go
+@@ -73,6 +73,7 @@ var flatOptions = map[string]bool{
+ 	"default-ulimits":    true,
+ 	"features":           true,
+ 	"builder":            true,
++	"base-runtime-spec":  true,
+ }
+
+ // skipValidateOptions contains configuration keys
+diff --git a/daemon/config/config_unix.go b/daemon/config/config_unix.go
+index 96805d3..1a16164 100644
+--- a/daemon/config/config_unix.go
++++ b/daemon/config/config_unix.go
+@@ -39,6 +39,7 @@ type Config struct {
+ 	NoNewPrivileges      bool                     `json:"no-new-privileges,omitempty"`
+ 	IpcMode              string                   `json:"default-ipc-mode,omitempty"`
+ 	CgroupNamespaceMode  string                   `json:"default-cgroupns-mode,omitempty"`
++	BaseRuntimeSpec      string                   `json:"base-runtime-spec,omitempty"`
+ 	// ResolvConf is the path to the configuration of the host resolver
+ 	ResolvConf string `json:"resolv-conf,omitempty"`
+ 	Rootless   bool   `json:"rootless,omitempty"`
+diff --git a/daemon/oci_linux.go b/daemon/oci_linux.go
+index a5a5acf..53908ce 100644
+--- a/daemon/oci_linux.go
++++ b/daemon/oci_linux.go
+@@ -2,6 +2,7 @@ package daemon // import "github.com/docker/docker/daemon"
+
+ import (
+ 	"context"
++	"encoding/json"
+ 	"fmt"
+ 	"io/ioutil"
+ 	"os"
+@@ -24,6 +25,7 @@ import (
+ 	"github.com/docker/docker/pkg/stringid"
+ 	"github.com/docker/docker/rootless/specconv"
+ 	volumemounts "github.com/docker/docker/volume/mounts"
++	units "github.com/docker/go-units"
+ 	"github.com/moby/sys/mount"
+ 	"github.com/moby/sys/mountinfo"
+ 	"github.com/opencontainers/runc/libcontainer/apparmor"
+@@ -45,8 +47,13 @@ func WithRlimits(daemon *Daemon, c *container.Container) coci.SpecOpts {
+
+ 		// We want to leave the original HostConfig alone so make a copy here
+ 		hostConfig := *c.HostConfig
++
++		//Merge defaults in Base Spec file with Daemon defaults
++		daemon.mergeBaseSpecUlimits(s)
++
+ 		// Merge with the daemon defaults
+ 		daemon.mergeUlimits(&hostConfig)
++
+ 		for _, ul := range hostConfig.Ulimits {
+ 			rlimits = append(rlimits, specs.POSIXRlimit{
+ 				Type: "RLIMIT_" + strings.ToUpper(ul.Name),
+@@ -158,8 +165,12 @@ func WithApparmor(c *container.Container) coci.SpecOpts {
+ // WithCapabilities sets the container's capabilties
+ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	return func(ctx context.Context, _ coci.Client, _ *containers.Container, s *coci.Spec) error {
++		defCaps := caps.DefaultCapabilities()
++		defCaps = append(defCaps, s.Process.Capabilities.Bounding...)
++		defCaps = getUniqueCapabilities(defCaps)
++
+ 		capabilities, err := caps.TweakCapabilities(
+-			caps.DefaultCapabilities(),
++			defCaps,
+ 			c.HostConfig.CapAdd,
+ 			c.HostConfig.CapDrop,
+ 			c.HostConfig.Privileged,
+@@ -171,6 +182,19 @@ func WithCapabilities(c *container.Container) coci.SpecOpts {
+ 	}
+ }
+
++func getUniqueCapabilities(defCaps []string) []string {
++	occurred := map[string]bool{}
```

Comment:
A potential optimization since you know there will never be more elements than what is in `defCaps` would be to make this:

```go
occurred := make(map[string]bool, len(defCaps))
```